### PR TITLE
feat(pipe): add component-resolved TwoFluidPipe transport

### LINKED ISSUE #2826: Add component-resolved transient transport to TwoFluidPipe
## Problem

The current `TwoFluidPipe` model conserves phase-level mass and momentum and preserves gas/oil/water phase identity, but it does not transport a full independent component composition in every cell and phase. The model documentation says this explicitly.

Therefore a CO2 pulse cannot be rigorously traced through a two-phase export pipeline: phase appearance, dissolution/evaporation, interphase transfer, and the resulting energy effects cannot conserve each chemical component through the grid.

Reviewed against NeqSim master commit `7ae08253b45180045a99a700314f23c9a7001b3c`.

## Proposed staged implementation

1. Add per-cell, per-phase component inventories for gas, oil, and water.
2. Advect component fluxes consistently with accepted phase mass fluxes.
3. Map interphase transfer by component identity and enforce equal/opposite source terms.
4. Synchronize cell thermodynamics without changing the conservative inventories.
5. Extend thermal accounting for compositional enthalpy and latent effects.
6. Add immutable per-step component, phase, and energy conservation reports and history.
7. Fail loudly for unsupported component slates, phase transitions, or boundary states.

A practical first gate can use two components and two phases before generalizing.

## Acceptance criteria

- One-phase limit reproduces the validated `PipeFlowSystem` pulse within stated discretization differences.
- A two-phase condensation/dissolution case conserves every component across both phases.
- Interphase component transfer is equal and opposite cell by cell.
- Total phase mass, component mass, and energy ledgers close to documented tolerances.
- Phase identity is preserved through flash synchronization.
- Positive/bounded phase fractions and compositions.
- Deterministic repeated run plus grid/timestep refinement.
- Java and Python-accessible diagnostics.
- Documentation states equations, closures, boundary conditions, and unsupported regimes.
- Java 8 and focused pre-commit/tests pass.

## Motivation

This is required before `TwoFluidPipe` can support dynamic gas-quality/CO2 tracking in wet-gas export systems and before it can be coupled credibly to transient multi-source pipeline networks.

### PR DESCRIPTION
**Pull request by EvenSol**
## Summary

- add opt-in per-cell, per-phase named-component inventories to `TwoFluidPipe`
- advect components with the exact accepted gas/oil/water face fluxes and integration-stage weights
- map flash-driven condensation/evaporation by component identity with equal-and-opposite cell sources
- rebuild cell PT flashes from conservative component inventories without projecting those inventories
- add phase-specific partial-component enthalpy and latent heat to the transient temperature equation and energy residual
- expose immutable component reports/history, bounded profiles, outlet composition, stable JSON, and JPype imports
- fail loudly for changed slates, unsupported phases/transitions, non-closing ledgers, negative inventories, reverse outlet inflow, and malformed public report inputs

## Baseline and bounded improvement

Before this work, `TwoFluidPipe` transported only aggregate gas/oil/water mass. It could not preserve named components through a transient or audit component identity across phase appearance. A separate API baseline also reproduced that the public component-report constructor accepted a two-row “phase” matrix and only failed later through an index/NPE path.

This PR keeps component transport disabled by default. The accepted hydrodynamic phase state remains authoritative; the new component state is conservative and must synchronize to it within tolerance. The public report constructor now rejects missing/duplicate names, mismatched component/phase/cell shapes, non-finite required values, and negative elapsed time/substep counts at the Java/JPype boundary.

## Equations, units, and closures

For cell `i`, phase `k`, component `c`, and accepted face `f`:

```text
M(i,k,c,n+1) = M(i,k,c,n)
             + dt * [F(i-1/2,k,c) - F(i+1/2,k,c)]
             + delta_M_transfer(i,k,c)

F(f,k,c) = m_dot(f,k) * Y(upwind(f,k),k,c)
```

`M` and `delta_M_transfer` are in kg, `dt` is in s, and `F` and `m_dot` are in kg/s. The phase mass flux and Euler/RK/IMEX stage weights are exactly those accepted by phase continuity. For every cell and component:

```text
sum(k in {gas, oil, water}) delta_M_transfer(i,k,c) = 0
```

Flash-driven evaporation withdraws donor-phase composition; condensation uses the receiving equilibrium-phase composition. Direct oil-water transfer is not inferred and fails loudly. The interphase energy term is:

```text
Q_latent(i) = -sum(k,c) delta_M_transfer(i,k,c) * h_partial(i,k,c)
```

`h_partial` is the phase-specific partial component enthalpy in J/kg evaluated at cell pressure, temperature, and conservative composition; `Q_latent` is in J. Positive `Q_latent` enters the sensible-energy equation.

Assumptions and initial validity range: fixed named slate with at least two components; gas, hydrocarbon-liquid, and aqueous identities; positive inlet and outlet flow or closed boundaries; signed internal-face reversal; first-order upwind advection under the hydrodynamic CFL step. Unsupported phase identity, direct oil-water exchange, changed slate, reverse outlet inflow, material projection, or failed conservation throws.

## Validation

Conflict-resolution validation after merging current `master` `f7f96cd635c093b6205e6e8d3569d1f5f98abfdb`:

- resolved overlapping changelog, JPype setup, and executable-documentation test additions by retaining both the TwoFluidPipe and transient gas-network APIs
- merged focused suite: 84 tests passed (component reports/transport, gas-network transport, column warm-state regression, and documentation examples)
- full TwoFluid component suite including slow physical regressions: 7 tests passed
- `python -m py_compile devtools/neqsim_dev_setup.py` passed
- `pre-commit run --all-files --hook-stage pre-commit` passed without modifications
- `pre-commit run --all-files --hook-stage pre-push` passed, including Spotless and the 646-page documentation-search audit

Local validation of the change set before publication:

- new public report contract: 3 tests passed; the shape regression fails on the previous head
- executable documentation gate: the exact component-mode setup, report, profile, outlet, and JSON calls compile and run in `DocExamplesCompilationTest`
- fast component transport/report suite: 7 tests passed
- slow component suite: 2 tests passed in 32.78 s
  - compact one-phase pulse compared with validated `PipeFlowSystem`; coarse and refined outlet responses stay within the documented 0.08 absolute mass-fraction envelope, with refinement not moving away by more than 0.01
  - closed SRK-CPA water-dew-point transition conserves every component across gas/oil/water, verifies equal-and-opposite transfer and non-zero latent heat, and closes phase and thermal ledgers
- Spotless apply/check passed
- `pre-commit run --all-files --hook-stage pre-commit` passed
- `pre-commit run --all-files --hook-stage pre-push` passed
- documentation-search audit passed for 646 Markdown pages and 1 standalone HTML page
- Python syntax and JPype end-to-end component transport/report smoke passed on the preceding implementation head; no Python-facing implementation changed in the report-validation update

Exact PR head: `ef9785d9dd08c56610b5727c0e0c61cebb0e827c`, with current `master` `f7f96cd635c093b6205e6e8d3569d1f5f98abfdb` merged as the second parent. Hosted Java 8/21, Windows, slow shards, Javadocs, static analysis, and review must pass on this head before merge readiness.

## Literature, public tools, and benchmark limits

Recent open literature shows why a bounded first-order baseline must not be represented as state-of-the-art front resolution: high-resolution/WENO composition tracking resolves sharp gas-mixture transients more accurately ([International Journal of Hydrogen Energy, 2024](https://doi.org/10.1016/j.ijhydene.2024.06.402)), while measured-boundary finite-difference work documents concentration sensitivity to numerical implicitness and reversal artifacts ([International Journal of Hydrogen Energy, 2024](https://doi.org/10.1016/j.ijhydene.2023.11.031)). A control-volume transient gas-network formulation provides an additional published conservation architecture reference ([Energy, 2021](https://doi.org/10.1016/j.energy.2020.119131)).

Public product documentation says OLGA supports transient compositional tracking and three-phase oil/water/gas flow ([OLGA overview](https://www.slb.com/products-and-services/delivering-digital-at-scale/software/olga), [OLGA core](https://www.slb.com/products-and-services/delivering-digital-at-scale/software/olga/olga-pipeline-management/olga-core-system)). LedaFlow 2.12 publicly reports custom-fluid tracking and a less numerically diffusive energy equation ([release note, 14 May 2025](https://kongsbergdigital.com/news/ledaflow-2.12-release)). These sources motivate gap prioritization only; this PR neither copies proprietary implementations nor claims simulator equivalence.

No public phase-resolved experimental component-transition dataset matching this geometry, fluid model, and closure set was found. The current evidence is therefore an independent NeqSim solver limit, a rigorous EOS phase-transition case, conservation, boundedness, determinism, and refinement trends—not experimental validation. The next numerical target is a separately scoped high-resolution/TVD or explicitly parameterized physical-dispersion model with a published pulse/step benchmark.

## Documentation and compatibility

Documentation impact: **updated**.

- `docs/process/TWOFLUIDPIPE_MODEL.md` defines equations, units, closures, boundary conditions, tolerance behavior, validation envelope, report validation, and unsupported regimes.
- Javadocs describe the public Java/JPype report contract and latent-inclusive thermal ledger.
- `devtools/neqsim_dev_setup.py` exposes the new classes and phase enum.
- `CHANGELOG_AGENT_NOTES.md` records the opt-in API and initial validated scope.
- Component transport remains disabled by default; existing public behavior is unchanged.

A one-line formatter-only adjustment to the newly merged ASTM D4052 regression is included because current-master Spotless rewrites that line during the mandatory all-files gate.

Closes #2826

### CODE REVIEW

File: src/main/java/neqsim/process/equipment/pipeline/TwoFluidPipe.java [hunk 1]
```diff
@@ -4788,6 +4879,124 @@ public TwoFluidThermalEnergyBalanceReport getLastThermalEnergyBalanceReport() {
     return lastThermalEnergyBalanceReport;
```
[the remark this answers is not in the document]
Comment (EvenSol) [reply src/main/java/neqsim/process/equipment/pipeline/TwoFluidPipe.java:4885]: Fixed in 2192c8c8. The getter Javadoc now documents the full thermal ledger, including component-resolved latent heat, and explains why component transport evaluates the thermal step even when external heat transfer is zero.

File: src/main/java/neqsim/process/equipment/pipeline/twophasepipe/TwoFluidComponentTransport.java [hunk 2]
```diff
@@ -0,0 +1,736 @@
...
+  public String[] getComponentNames() {
+    return Arrays.copyOf(componentNames, componentNames.length);
+  }
+
+  private double[][] allocateComponentTransfer(int cell, double[] phaseTransferKg, double[][][] updated,
+      TwoFluidSection[] sections, SystemInterface fluidTemplate) {
+    double[][] transfer = new double[PHASE_COUNT][componentNames.length];
+    double gasTransfer = phaseTransferKg[GAS];
+    if (gasTransfer > 0.0) {
+      for (int donorPhase : new int[] { OIL, WATER }) {
```
[the remark this answers is not in the document]
Comment (EvenSol) [reply src/main/java/neqsim/process/equipment/pipeline/twophasepipe/TwoFluidComponentTransport.java:346]: Fixed in 2192c8c8. Opposite-signed oil/water transfers are rejected before any component allocation, including when gas transfer is nonzero. A focused regression exercises `[gas, oil, water] = [1, -2, 1]` and verifies the fail-loud diagnostic.

**Comment by EvenSol** [comment]
Exact-head engineering audit: `4ea8b0e1b5e4a18a5634527caa6743855261540c`

- Rebased directly on current `master` `39c690c49261ba5b092fe3576781d9d57b29a1ef`; compare is one ahead / zero behind and GitHub reports mergeable.
- Added fail-loud validation for the public report constructor after reproducing acceptance of a two-row phase matrix. Three new contract tests pass; the malformed-shape regression fails on the preceding head.
- Re-ran 7 fast component/report tests and both slow component cases (32.78 s), plus both mandatory pre-commit stages and documentation search.
- All review threads remain resolved. Copilot reviewed all 12 changed files on this head and produced no new review comments.
- The suppressed suggestion to split `prepareFluid()` initialization was technically assessed and is intentionally not included: the helper also serves `createThermodynamicState()`, whose caller consumes density, viscosity, and sound speed, while the caloric-only paths depend on complete phase initialization. No repeatable speed baseline or equivalence proof supports changing that lifecycle in this scientific PR; no speed claim is made.
- Hosted Devtools, PaperLab, pre-commit, Spotless, documentation search, and CodeQL pass. Java build/test/Javadocs and slow shards are still running.

Prioritized remaining gap map after this foundational component state:
1. Published step/pulse benchmark plus high-resolution/TVD numerical advection or explicitly parameterized physical dispersion.
2. Validated reverse/mixed boundary compositions and startup/shutdown handling.
3. Direct oil-water transfer and fully coupled three-phase closure.
4. Terrain/flow-regime-transition benchmarks, then hydrodynamic/severe slugging and minimum-stable-flow cases.

The literature and public OLGA/LedaFlow capability context, limitations, equations, units, and documentation impact are recorded in the PR description. No proprietary implementation or simulator-equivalence claim is used.

**Comment by EvenSol** [comment]
Updated exact head: `5470bf73580d4ca559ec954b2004e680d9125c69`.

The latest suppressed documentation-review suggestion was valid and is fixed. `DocExamplesCompilationTest.testTwoFluidPipeSteadyAndDynamicBoundaryDocExamples()` now runs the documented component-mode setup and exercises the report, gas-profile, outlet-composition, and JSON APIs using a CO2-bearing fluid. Focused result: 4 tests passed (3 report-contract tests plus the executable documentation test). Spotless, both mandatory pre-commit stages, and the 646-Markdown/1-HTML documentation-search audit pass after the change.

This push contains no physics, solver, tolerance, or public-API change. Hosted exact-head CI and review have restarted; prior-head check results are not used for merge readiness.

**Comment by EvenSol** [comment]
Addressed the final boundary-contract audit on exact head `bc875dd51a1d17b6733f3d675cd23e411b4e178d`.

Negative flow at face 0 now fails loudly as unsupported inlet-boundary outflow, matching the documented positive-inlet-flow scope. The obsolete path that recorded this mass as outlet flow was removed, while signed internal-face reversal remains supported. Added `reverseFlowThroughInletFailsBeforeComponentAdvection()` as a focused regression.

Local verification: the new test passes and `spotless:check` passes. Hosted exact-head CI is restarting; earlier-head results will not be used for merge readiness.

**Comment by EvenSol** [comment]
Final fail-loud contract fixes are published on exact head `ecc5db1de8531f56a751d54e91de3aba33520982`.

- `TwoFluidPipe.runTransient(...)` now rejects zero, negative, NaN, and infinite durations before mutating reports or component history.
- `TwoFluidComponentTransport.advance(...)` and `createThermodynamicState(...)` now reject a null fluid template with explicit `IllegalArgumentException`s.
- Updated the thermal-step inline comment to match component-enabled behavior.
- Added focused regressions for duration/history invariants and both null-template entry points.

Local verification: all three focused fail-loud regressions pass together; `spotless:check` passes. Hosted exact-head CI is restarting, and prior-head build results are discarded.
